sky/adaptors/azure.py

```diff
@@ -55,10 +62,123 @@ def get_client(name: str, subscription_id: str):
+            from azure.mgmt.authorization import AuthorizationManagementClient
+            return AuthorizationManagementClient(credential, subscription_id)
+        elif name == 'graph':
+            from msgraph import GraphServiceClient
+            return GraphServiceClient(credential)
+        elif name == 'container':
+            from azure.storage.blob import ContainerClient
+            container_url = kwargs.pop('container_url', None)
+            assert container_url is not None
+            container_client = ContainerClient.from_container_url(container_url)
+            try:
+                container_client.exists()
+            except exceptions().ClientAuthenticationError as e:
+                # Caught when credential is not provided to the private
+                # container url or wrong container name is provided as the
+                # public container url. We reattempt with credentials assuming
+                # user is using private container with access rights.
+                if 'ErrorCode:NoAuthenticationInformation' in e.message:
+                    container_client = ContainerClient.from_container_url(
+                        container_url, credential)
+                    try:
+                        # Suppress noisy logs from Azure SDK when attempting
+                        # to run exists() on public container with credentials.
+                        # Reference:
+                        # https://github.com/Azure/azure-sdk-for-python/issues/9422 # pylint: disable=line-too-long
+                        azure_logger = logging.getLogger('azure')
+                        original_level = azure_logger.getEffectiveLevel()
+                        azure_logger.setLevel(logging.CRITICAL)
+                        container_client.exists()
+                        azure_logger.setLevel(original_level)
+                    except exceptions().ClientAuthenticationError as error:
+                        # Caught when user attempted to use private container
+                        # without access rights.
+                        # Reference:
+                        # https://learn.microsoft.com/en-us/troubleshoot/azure/entra/entra-id/app-integration/error-code-aadsts50020-user-account-identity-provider-does-not-exist # pylint: disable=line-too-long
+                        if 'ERROR: AADSTS50020' in error.message:
+                            with ux_utils.print_exception_no_traceback():
+                                raise sky_exceptions.StorageBucketGetError(
+                                    'Attempted to fetch a non-existant public '
+                                    'container name: '
+                                    f'{container_client.container_name}. '
+                                    'Please check if the name is correct.')
+            return container_client
         else:
             raise ValueError(f'Client not supported: "{name}"')
 
 
+@functools.lru_cache()
```

Comment:
should we use lru cache here since the sas token generated is volatile and will be expired within an hour? same for the function below

Reply:
Great catch! I also measured the runtime of this function to retrieve the sas token, and it was fairly quick (0.0002732276916503906 sec). I removed the lru cache decorator.